### ISSUE TITLE
END TO END SALES ANALYTICS

### ISSUE THREAD

**Issue by SINGIPURAPUDINESH** [SINGIPURAPUDINESH/PYTHON-INTERNSHIP-PROJECTS#3]
<html>
<body>
<html><head></head><body><h1>📊 End-to-End Sales Analytics &amp; Prediction System</h1><p>An end-to-end data analytics and machine learning application that transforms raw sales data into actionable business insights and future sales forecasts.</p><p>The project demonstrates a complete data workflow:</p><p><strong>Data Cleaning → MySQL Database → SQL Analytics → KPI Analysis → Interactive Visualization → Machine Learning Prediction → Streamlit Dashboard</strong></p><hr><h2>🚀 Project Overview</h2><p>The <strong>Sales Analytics &amp; Prediction System</strong> is designed to analyze historical sales performance and estimate future sales trends.</p><p>The system takes raw transactional sales data, cleans and preprocesses it using Python, stores the processed data in MySQL, performs analytical operations, and presents the results through an interactive Streamlit dashboard.</p><p>A machine learning model is also used to forecast future monthly sales based on historical sales patterns.</p><hr><h2>🎯 Objectives</h2><ul><li><p>Clean and preprocess raw sales transaction data.</p></li><li><p>Remove duplicates and handle invalid or missing values.</p></li><li><p>Calculate revenue after applying discounts.</p></li><li><p>Store structured sales data in MySQL.</p></li><li><p>Analyze key business performance indicators.</p></li><li><p>Identify top-performing products and customers.</p></li><li><p>Compare sales performance across regions and categories.</p></li><li><p>Visualize historical sales trends interactively.</p></li><li><p>Forecast future monthly sales using machine learning.</p></li><li><p>Provide an easy-to-use business analytics dashboard.</p></li></ul><hr><h2>🛠️ Tech Stack</h2>
Technology | Purpose
-- | --
Python | Data processing and application logic
Pandas | Data cleaning and transformation
NumPy | Numerical computations
MySQL | Relational database storage
Scikit-learn | Machine learning and sales prediction
Plotly | Interactive data visualization
Streamlit | Interactive web dashboard
python-dotenv | Environment variable management
Git & GitHub | Version control and project hosting

<hr><h2>🏗️ System Architecture</h2><pre><code class="language-text">                    ┌─────────────────────┐
                    │   Raw Sales CSV     │
                    └──────────┬──────────┘
                               │
                               ▼
                    ┌─────────────────────┐
                    │  Data Cleaning      │
                    │  Python + Pandas    │
                    └──────────┬──────────┘
                               │
                               ▼
                    ┌─────────────────────┐
                    │  Cleaned Sales Data │
                    └──────────┬──────────┘
                               │
                               ▼
                    ┌─────────────────────┐
                    │       MySQL         │
                    │      Database       │
                    └──────────┬──────────┘
                               │
                 ┌─────────────┴─────────────┐
                 ▼                           ▼
       ┌─────────────────┐        ┌──────────────────┐
       │ Business        │        │ Machine Learning │
       │ Analytics       │        │ Sales Prediction │
       └────────┬────────┘        └─────────┬────────┘
                │                           │
                └─────────────┬─────────────┘
                              ▼
                   ┌─────────────────────┐
                   │ Streamlit Dashboard │
                   └─────────────────────┘
</code></pre><hr><h2>📁 Project Structure</h2><pre><code class="language-text">sales-analytics-prediction/
│
├── data/
│   ├── sales_data.csv
│   └── cleaned_sales_data.csv
│
├── sql/
│   ├── database.sql
│   └── analysis_queries.sql
│
├── src/
│   ├── __init__.py
│   ├── data_cleaning.py
│   ├── database.py
│   ├── analytics.py
│   └── prediction.py
│
├── app.py
├── requirements.txt
├── .env
├── .gitignore
└── README.md
</code></pre><hr><h1>🔄 Project Workflow</h1><h2>1. Data Ingestion</h2><p>The system starts with raw sales transaction data stored in CSV format.</p><p>Example fields include:</p><ul><li><p>Order ID</p></li><li><p>Order Date</p></li><li><p>Customer ID</p></li><li><p>Product</p></li><li><p>Category</p></li><li><p>Region</p></li><li><p>Quantity</p></li><li><p>Unit Price</p></li><li><p>Discount</p></li></ul><hr><h2>2. Data Cleaning</h2><p>The <code inline="">data_cleaning.py</code> module performs preprocessing operations including:</p><ul><li><p>Duplicate removal</p></li><li><p>Date conversion</p></li><li><p>Numerical type conversion</p></li><li><p>Missing-value handling</p></li><li><p>Invalid record removal</p></li><li><p>Discount validation</p></li><li><p>Revenue calculation</p></li><li><p>Date feature extraction</p></li></ul><h3>Sales Calculation</h3><p>Revenue is calculated using:</p><pre><code class="language-text">Sales = Quantity × Unit Price × (1 − Discount)
</code></pre><p>Additional time-based features include:</p><ul><li><p>Year</p></li><li><p>Month</p></li><li><p>Month Name</p></li><li><p>Quarter</p></li></ul><hr><h2>3. Database Storage</h2><p>The cleaned data is stored in a MySQL database.</p><p>Database:</p><pre><code class="language-text">sales_analytics
</code></pre><p>Table:</p><pre><code class="language-text">sales
</code></pre><p>The database layer handles:</p><ul><li><p>MySQL connection</p></li><li><p>Table creation</p></li><li><p>Data insertion</p></li><li><p>Duplicate order handling</p></li><li><p>Data retrieval</p></li></ul><hr><h1>📊 Business Analytics</h1><p>The project calculates several important business KPIs.</p><h3>Key Performance Indicators</h3><ul><li><p>Total Sales</p></li><li><p>Total Orders</p></li><li><p>Total Customers</p></li><li><p>Total Quantity Sold</p></li><li><p>Average Order Value</p></li></ul><h3>Sales Analysis</h3><p>The dashboard provides analysis of:</p><ul><li><p>Monthly sales trends</p></li><li><p>Product performance</p></li><li><p>Category performance</p></li><li><p>Regional performance</p></li><li><p>Top customers</p></li><li><p>Product quantity sold</p></li></ul><hr><h1>📈 Interactive Dashboard</h1><p>The Streamlit dashboard provides interactive filtering and visualization.</p><p>Users can filter sales data by:</p><ul><li><p>Date range</p></li><li><p>Category</p></li><li><p>Region</p></li></ul><p>The dashboard dynamically updates the KPIs and visualizations based on the selected filters.</p><h3>Dashboard Components</h3><ul><li><p>📊 KPI cards</p></li><li><p>📈 Monthly sales trend</p></li><li><p>🏆 Product revenue analysis</p></li><li><p>📦 Category revenue distribution</p></li><li><p>🌎 Regional sales performance</p></li><li><p>👥 Top customer analysis</p></li><li><p>📦 Product quantity analysis</p></li><li><p>🔮 Future sales prediction</p></li><li><p>🤖 Model performance metrics</p></li><li><p>📄 Filtered data table</p></li><li><p>⬇️ CSV data download</p></li></ul><hr><h1>🤖 Machine Learning</h1><p>The project includes a sales forecasting component using <strong>Scikit-learn Linear Regression</strong>.</p><p>Historical sales are aggregated at the monthly level.</p><h3>Features</h3><p>The model uses:</p><ul><li><p>Month number</p></li><li><p>Year</p></li><li><p>Month</p></li></ul><h3>Target</h3><pre><code class="language-text">Monthly Sales
</code></pre><p>The dataset is divided into chronological training and testing sets to avoid randomly mixing historical and future observations.</p><hr><h2>📐 Model Evaluation</h2><p>The model is evaluated using:</p><h3>Mean Absolute Error (MAE)</h3><pre><code class="language-text">MAE = average(|Actual − Predicted|)
</code></pre><h3>Root Mean Squared Error (RMSE)</h3><pre><code class="language-text">RMSE = √MSE
</code></pre><h3>R² Score</h3><p>R² measures how well the model explains the variation in the target variable.</p><p>The dashboard displays:</p><ul><li><p>MAE</p></li><li><p>RMSE</p></li><li><p>R² Score</p></li></ul><hr><h1>🔮 Sales Forecasting</h1><p>The application allows users to select the number of future months to forecast.</p><p>Supported forecast horizon:</p><pre><code class="language-text">1–12 months
</code></pre><p>The dashboard displays:</p><ul><li><p>Historical sales</p></li><li><p>Predicted sales</p></li><li><p>Forecast table</p></li><li><p>Historical vs predicted sales visualization</p></li></ul><hr><h1>💻 Installation</h1><h2>Prerequisites</h2><p>Make sure you have the following installed:</p><ul><li><p>Python 3.9+</p></li><li><p>MySQL 8.0+</p></li><li><p>Git</p></li></ul><hr><h2>1. Clone the Repository</h2><pre><code class="language-bash">git clone https://github.com/YOUR_USERNAME/sales-analytics-prediction.git
</code></pre><p>Navigate to the project:</p><pre><code class="language-bash">cd sales-analytics-prediction
</code></pre><hr><h2>2. Create a Virtual Environment</h2><h3>Windows</h3><pre><code class="language-bash">python -m venv venv
</code></pre><p>Activate it:</p><pre><code class="language-bash">venv\Scripts\activate
</code></pre><h3>macOS/Linux</h3><pre><code class="language-bash">python3 -m venv venv
</code></pre><p>Activate it:</p><pre><code class="language-bash">source venv/bin/activate
</code></pre><hr><h2>3. Install Dependencies</h2><pre><code class="language-bash">pip install -r requirements.txt
</code></pre><hr><h1>🗄️ MySQL Configuration</h1><p>Create the database using the SQL script.</p><p>Open MySQL and run:</p><pre><code class="language-sql">CREATE DATABASE IF NOT EXISTS sales_analytics;

USE sales_analytics;

CREATE TABLE IF NOT EXISTS sales (
    Order_ID INT PRIMARY KEY,
    Order_Date DATE,
    Customer_ID VARCHAR(50),
    Product VARCHAR(100),
    Category VARCHAR(100),
    Region VARCHAR(50),
    Quantity INT,
    Unit_Price DECIMAL(12,2),
    Discount DECIMAL(5,2),
    Sales DECIMAL(15,2)
);
</code></pre><hr><h1>🔐 Environment Variables</h1><p>Create a <code inline="">.env</code> file in the project root:</p><pre><code class="language-env">DB_HOST=localhost
DB_USER=root
DB_PASSWORD=your_mysql_password
DB_NAME=sales_analytics
</code></pre><h3>Important</h3><p>Never commit your <code inline="">.env</code> file to GitHub.</p><p>Your <code inline="">.gitignore</code> should contain:</p><pre><code class="language-gitignore">.env
__pycache__/
*.pyc
venv/
.venv/
models/
</code></pre><hr><h1>▶️ Running the Project</h1><p>Run the components in the following order.</p><h3>Step 1 — Clean the data</h3><pre><code class="language-bash">python src/data_cleaning.py
</code></pre><p>Expected output:</p><pre><code class="language-text">Data cleaning completed.
Rows before cleaning: 20
Rows after cleaning: 20
</code></pre><hr><h3>Step 2 — Load data into MySQL</h3><pre><code class="language-bash">python src/database.py
</code></pre><p>Expected output:</p><pre><code class="language-text">Data successfully loaded into MySQL.
</code></pre><hr><h3>Step 3 — Start the Streamlit dashboard</h3><pre><code class="language-bash">python -m streamlit run app.py
</code></pre><p>The application will be available at:</p><pre><code class="language-text">http://localhost:8501
</code></pre><hr><h1>📋 SQL Analytics</h1><p>The project includes predefined SQL queries in:</p><pre><code class="language-text">sql/analysis_queries.sql
</code></pre><p>These queries cover:</p><ul><li><p>Total sales</p></li><li><p>Total orders</p></li><li><p>Total customers</p></li><li><p>Average order value</p></li><li><p>Monthly sales</p></li><li><p>Sales by product</p></li><li><p>Sales by category</p></li><li><p>Sales by region</p></li><li><p>Top customers</p></li><li><p>Product quantity</p></li><li><p>Average discount</p></li><li><p>Highest-value order</p></li><li><p>Category-region performance</p></li></ul><hr><h1>🧪 Data Quality Checks</h1><p>The preprocessing pipeline performs several validation steps:</p><pre><code class="language-text">Raw Data
   │
   ├── Duplicate Check
   │
   ├── Date Validation
   │
   ├── Numeric Conversion
   │
   ├── Missing Value Handling
   │
   ├── Quantity Validation
   │
   ├── Price Validation
   │
   ├── Discount Validation
   │
   └── Revenue Calculation
</code></pre><hr><h1>📌 Current Limitations</h1><p>The current version uses a relatively small sample dataset and a simple Linear Regression model.</p><p>Because sales forecasting is a time-series problem, the model can be improved significantly with a larger real-world dataset.</p><p>Potential improvements include:</p><ul><li><p>More historical data</p></li><li><p>Lag features</p></li><li><p>Rolling averages</p></li><li><p>Seasonality</p></li><li><p>Holiday effects</p></li><li><p>Promotional campaigns</p></li><li><p>Product-level forecasting</p></li><li><p>Customer-level forecasting</p></li><li><p>Random Forest</p></li><li><p>Gradient Boosting</p></li><li><p>XGBoost</p></li><li><p>Prophet</p></li><li><p>ARIMA/SARIMA</p></li><li><p>Cross-validation for time-series data</p></li><li><p>Prediction confidence intervals</p></li></ul><hr><h1>🧠 Skills Demonstrated</h1><p>This project demonstrates practical skills in:</p><h3>Data Analytics</h3><ul><li><p>Data cleaning</p></li><li><p>Exploratory data analysis</p></li><li><p>KPI development</p></li><li><p>Business intelligence</p></li><li><p>Data visualization</p></li><li><p>SQL analytics</p></li></ul><h3>Python</h3><ul><li><p>Pandas</p></li><li><p>NumPy</p></li><li><p>Modular programming</p></li><li><p>Data transformation</p></li><li><p>Database integration</p></li></ul><h3>SQL</h3><ul><li><p>Database design</p></li><li><p>Table creation</p></li><li><p>Aggregations</p></li><li><p>GROUP BY</p></li><li><p>ORDER BY</p></li><li><p>Date functions</p></li><li><p>Business analysis queries</p></li></ul><h3>Machine Learning</h3><ul><li><p>Feature engineering</p></li><li><p>Regression</p></li><li><p>Train/test splitting</p></li><li><p>Model evaluation</p></li><li><p>Forecasting</p></li></ul><h3>Visualization</h3><ul><li><p>Plotly</p></li><li><p>Interactive charts</p></li><li><p>Dashboard development</p></li></ul><h3>Application Development</h3><ul><li><p>Streamlit</p></li><li><p>Environment configuration</p></li><li><p>End-to-end data pipeline</p></li></ul><hr><h1>📷 Dashboard Preview</h1><p>Add screenshots of your Streamlit dashboard here after running the application.</p><p>Example:</p><pre><code class="language-text">docs/
└── dashboard.png
</code></pre><p>Then add:</p><pre><code class="language-markdown">![Sales Analytics Dashboard](docs/dashboard.png)
</code></pre><hr><h1>📈 Project Pipeline</h1><pre><code class="language-text">                    RAW DATA
                       │
                       ▼
              ┌─────────────────┐
              │ Data Cleaning   │
              │    Pandas       │
              └────────┬────────┘
                       │
                       ▼
              ┌─────────────────┐
              │    MySQL DB     │
              └────────┬────────┘
                       │
             ┌─────────┴─────────┐
             │                   │
             ▼                   ▼
      ┌─────────────┐     ┌──────────────┐
      │  Analytics  │     │ ML Forecast  │
      └──────┬──────┘     └──────┬───────┘
             │                   │
             └─────────┬─────────┘
                       ▼
              ┌─────────────────┐
              │   Streamlit     │
              │    Dashboard    │
              └─────────────────┘
</code></pre><hr><h1>⭐ Project Highlights</h1><ul><li><p>End-to-end data analytics pipeline</p></li><li><p>Python-based data preprocessing</p></li><li><p>MySQL database integration</p></li><li><p>SQL business analytics</p></li><li><p>Interactive Streamlit dashboard</p></li><li><p>Plotly visualizations</p></li><li><p>Machine learning sales forecasting</p></li><li><p>Model evaluation using MAE, RMSE, and R²</p></li><li><p>Interactive filters</p></li><li><p>Downloadable cleaned sales data</p></li></ul></body></html>
</body>
</html>

[analytics.py](https://github.com/user-attachments/files/31119859/analytics.py)
[data_cleaning.py](https://github.com/user-attachments/files/31119862/data_cleaning.py)
[database.py](https://github.com/user-attachments/files/31119860/database.py)
[prediction.py](https://github.com/user-attachments/files/31119861/prediction.py)
[app.py](https://github.com/user-attachments/files/31119864/app.py)
